pylint fixes for vault related code

alikins opened the pull request:
##### SUMMARY
A set of fixes for issues found by pylint, mostly for vault related code.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (vault_pylint 2eef044918) last updated 2017/08/03 17:29:56 (GMT -400)
  config file = /home/adrian/src/ansible/ansible.cfg
  configured module search path = [u'/home/adrian/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/adrian/src/ansible/lib/ansible
  executable location = /home/adrian/src/ansible/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]


```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

alikins committed:
rm unneeded parens following assert

pylint: Unnecessary parens after 'assert' keyword

alikins committed:
import six.zip as _zip to avoid shadowed builtin

pylint: Redefining built-in 'zip' (redefined-builtin)

alikins committed:
Add super().__init__ call to PromptVaultSecret.__init__

pylint: __init__ method from base class 'VaultSecret' is not called (super-init-not-called)

alikins committed:
rm unnecessary else

pylint: Unnecessary "else" after "return" (no-else-return)

alikins committed:
Make FileVaultSecret.read_file reg method again

The base class read_file() doesnt need self but
the sub classes do.

Rm now unneeded loader arg to read_file()

pylint: Unnecessary "else" after "return" (no-else-return)

alikins committed:
Fix err msg string literal that had no effect

The indent on the continuation of the msg_format was wrong
so the second half was dropped.

There was also no need to join() filename (copy/paste from
original with a command list I assume...)

pylint: String statement has no effect

alikins committed:
Use _ for some unused vars from tuple expansion

pylint: Unused variable 'stderr'
(etc)

alikins committed:
Use _ for unused b_version var

pylint: Unused variable 'b_version'

alikins committed:
Use local cipher_name in VaultEditor.edit_file not instance

Use the local cipher_name returned from parse_vaulttext_envelope()
instead of the instance self.cipher_name var.

Since there is only one valid cipher_name either way, it was
equilivent, but it will not be with more valid cipher_names

pylint: Unused variable 'cipher_name'

alikins committed:
Use _ for unused vars from tuple

We only use the cipher_name and vault_id here.

TODO: just make parse_vaulttext_envelope() return an object
or dict instead of a tuple

pylint: Unused variable 'b_ciphertext'

alikins committed:
Use equality of b_next_chunk instead of len()

b_next_chunk is a byte string, so True if not empty.

pylint: Do not use `len(SEQUENCE)` as condition value

alikins committed:
undo that last commit SQUASH ME

alikins committed:
rm unused vault_id arg on VaultAES.decrypt()

pylint: Unused argument 'vault_id'

alikins committed:
Rm unused b_salt arg on VaultAES256._encrypt*

Previously the methods computed the keys and iv themselves
so needed to be passed in the salt, but now the key/iv
are built before and passed in so b_salt arg is not used
anymore.

pylint: Unused argument 'b_salt'

alikins committed:
rm redundant import of call from subprocess

use via subprocess module now instead of direct
import.

pylint: Imports from package subprocess are not grouped

alikins committed:
rm unused VaultCli.secrets

pylint: Attribute 'editor' defined outside __init__

alikins committed:
No longer need index/enumerate over vault_ids

pylint: Unused variable 'index'

alikins committed:
rm unused parse_vaulttext_envelope from yaml.constructor

pylint: Unused parse_vaulttext_envelope imported from ansible.parsing.vault

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -344,23 +343,21 @@ def __repr__(self):
             raise AnsibleError(msg)
 
-        stdout, stderr = p.communicate()
+        stdout, _ = p.communicate()
```
we've avoided using  _ as a 'unused var'  in case we decide to use i18n .. which also has special uses for _

alikins replied:
Whats our preferred dummy var name?

alikins replied:
The default pylint match for dummy vars is:

```
dummy-variables-rgx=_+$|(_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy|^ignored_|^unused_
```

alikins replied:
```
newswoop:F25:ansible (vault_pylint %)]$ ack --python ' _+ ' | wc -l
82
```

the (_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)  regex hits _anything, private methods etc ~7000 hits, afaict none of them to indicate unused vars

'ignored' doesnt seem to be used much to indicate unused vars but is in a lot of var names

'unused_' gets no matches.


So of the defaults, _ is the only one we use, and we've used it 82 times, most to indicate unused/dummy vars (including a view usages of 'for _ in range(some_count)'.

bcoca replied:
i've seen plenty of 'dummies', i just have not seen a pattern to them.

alikins replied:
I see 0 uses of 'dummy' for that purpose in current code (in 'ack --python "dummy"' output)

abadger commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -75,7 +74,7 @@
 from ansible import constants as C
 from ansible.module_utils.six import PY3, binary_type
-from ansible.module_utils.six.moves import zip
+from ansible.module_utils.six.moves import zip as _zip
```
We should disable pylint's check for this one line rather than doing this.  doing this defeats part of the purpose of having a compatibility library.

alikins replied:
I don't mind it myself because the 'zip' I have imported is (sometimes) not the builtin 'zip', so the different name makes sense. Also generally dislike shadowing so I'd like to keep it for the general case.

But granted, 'six' exists to abstract those details...

pylint has a built in 'redefining-builtins-modules' option that by default includes 'six.moves,future.builtins' so should be able to add 'ansible.module_utils.six.moves' to the exceptions.

abadger replied:
I just checked and six.moves.zip always comes from Python.  In Python2, it's in builtins (and thus available globally) whereas in python3 it's from itertools.izip.

If you can figure out the invocation to add ansible.module_utils.six.moves that sounds good.  There's currently not an ansible pylint config file but you can modify the command line invocation of pylint here: https://github.com/ansible/ansible/blob/devel/test/runner/lib/sanity.py#L466

alikins replied:
re  py2/py3 zip, I think that is vice-versa. On py2 six.moves.zip is itertools.izip, and on py3 it is the builtin zip.

py2.7
```
[newswoop:F25:ansible (vault_pylint %)]$ python
Python 2.7.13 (default, May 10 2017, 20:04:28) 
[GCC 6.3.1 20161221 (Red Hat 6.3.1-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> print(zip)
<built-in function zip>
>>> print(zip.__doc__)
zip(seq1 [, seq2 [...]]) -> [(seq1[0], seq2[0] ...), (...)]

Return a list of tuples, where each tuple contains the i-th element
from each of the argument sequences.  The returned list is truncated
in length to the length of the shortest argument sequence.
>>> from ansible.module_utils.six.moves import zip
>>> print zip
<type 'itertools.izip'>
>>> print(zip.__doc__)
izip(iter1 [,iter2 [...]]) --> izip object

Return a izip object whose .next() method returns a tuple where
the i-th element comes from the i-th iterable argument.  The .next()
method continues until the shortest iterable in the argument sequence
is exhausted and then it raises StopIteration.  Works like the zip()
function but consumes less memory by returning an iterator instead of
a list.
```

py3.6
```
[newswoop:F25:ansible (vault_pylint %)]$ python3.6
Python 3.6.2 (default, Jul 24 2017, 14:22:34) 
[GCC 6.3.1 20161221 (Red Hat 6.3.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> print(zip)
<class 'zip'>
>>> print(zip.__doc__)
zip(iter1 [,iter2 [...]]) --> zip object

Return a zip object whose .__next__() method returns a tuple where
the i-th element comes from the i-th iterable argument.  The .__next__()
method continues until the shortest iterable in the argument sequence
is exhausted and then it raises StopIteration.
>>> from ansible.module_utils.six.moves import zip
>>> print(zip)
<class 'zip'>
>>> print(zip.__doc__)
zip(iter1 [,iter2 [...]]) --> zip object

Return a zip object whose .__next__() method returns a tuple where
the i-th element comes from the i-th iterable argument.  The .__next__()
method continues until the shortest iterable in the argument sequence
is exhausted and then it raises StopIteration.
```

abadger commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -240,6 +239,7 @@ class PromptVaultSecret(VaultSecret):
 
     def __init__(self, _bytes=None, vault_id=None, prompt_formats=None):
+        super(PromptVaultSecret, self).__init__(_bytes=_bytes)
         self._bytes = _bytes
```
Now that we're calling super, we should remove setting self._bytes in the PromptVaultSecret class, shouldn't we?

abadger commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -319,10 +319,9 @@ def bytes(self):
 
-    @staticmethod
-    def read_file(filename, loader):
+    def read_file(self, filename):
```
This is a bad change.  self is not used inside of this method at all.  I'm not sure which pylint warning you're seeing but the answer for it is probably to make this a toplevel function, private to this python module, not a regular method.

abadger replied:
Ah I see... but later you want to make this overridable by subclasses....  subclasses should be able to override a staticmethod with a regular method without difficulty so I think this should remain a method but also remain a staticmethod.  Depending on what the pylint warning is you can disable it here with a comment that this is a method because the method is overridden by subclasses.

alikins replied:
Nothing uses the read_file() statically so it doesn't gain anything other to make it static in one class and regular in another. I'd prefer them to be consistent.

But looking at it closer, it also doesnt need to be a public method at all. It is just a helper implementation of the the more abstract load().

Can probably make FileVaultSecret.read_file() use self.loader now (or possibly with
all or part of https://github.com/ansible/ansible/pull/27668/files and https://github.com/ansible/ansible/pull/27669/). So vault password files could use loader for encrypted files and whatever else dataloader enables.

abadger replied:
+1 to that I think.

alikins committed:
self._bytes is set in super init now, rm dup

alikins committed:
use 'dummy' for unused variables instead of _

Based on pylint unused variable warnings.

Existing code use '_' for this, but that is old
and busted. The hot new thing is 'dummy'. It
is so fetch.

Except for where we get warnings for reusing
the 'dummy' var name inside of a list comprehension.

alikins committed:
rm zip import as _zip

alikins committed:
Make FileVaultSecret.read_file() -> _read_file()

_read_file() is details of the implementation of
load(), so now 'private'.
